test/test_gpu_tensorcore.cpp

```diff
@@ -4045,6 +4045,69 @@ TEST_F(NVFuserTest, FusionAmpereMMATNAlpha_CUDA) {
   TORCH_CHECK(cg_outputs[0].allclose(tref, 0.0001, 0.0001));
 }
 
+TEST_F(NVFuserTest, FusionAmpereMatmulEpilogue_CUDA) {
+  // Keep multiples of 8 to keep vectorizable.
+  int M = 504, N = 136, K = 248;
+
+  for (auto layout : kAllSupportedMatmulLayout) {
+    Fusion fusion;
+    FusionGuard fg(&fusion);
+    auto tv0 = makeContigTensor(2, DataType::Half);
+    auto tv1 = makeContigTensor(2, DataType::Half);
+
+    fusion.addInput(tv0);
+    fusion.addInput(tv1);
+
+    auto tv2 = matmul(tv0, tv1, layout, true);
+
+    fusion.addOutput(tv2);
+
+    MatMulTileOptions gemm_tile;
+    gemm_tile.cta_tile = GemmTile(128, 128, 32);
+    gemm_tile.warp_tile = GemmTile(64, 64, 32);
+    gemm_tile.instruction_tile = GemmTile(16, 8, 16);
+
+    MatmulParams params;
+    params.mma_macro = MmaOptions::MacroType::Ampere_16_8_16;
+    params.tile_sizes = gemm_tile;
+    params.has_smem_epilogue = true;
+    params.async_gmem_load_operands = true;
+    params.double_buffer_options.double_buffer_smem_write = true;
+    params.double_buffer_options.double_buffer_smem_read = true;
+    params.double_buffer_options.smem_double_buffer_stage = 4;
+    scheduleMatmul(&fusion, params);
+
+    at::manual_seed(0);
+    auto inputs = matmulAtInput(M, N, K, layout);
+
+    FusionExecutor fe;
+    NVFUSER_TEST_CUDA_ARCH_COMPILE_CHECK(
+        8,
+        0,
+        fe.compileFusion(
+            &fusion,
+            {inputs.first, inputs.second},
+            LaunchParams(),
+            matmul_cparams));
+    auto cg_outputs = fe.runFusion({inputs.first, inputs.second});
+    auto tref = atMatmul(
+        inputs.first.to(at::kFloat), inputs.second.to(at::kFloat), layout);
+
+    // check bank conflicts
+    const auto& bank_conflict = getBankConflictInfo(fe.kernel());
```

Comment:
I wonder if scenario, where no epilogue shared memory tensor is generate, will also pass this check. My thinking is that there won't be any bank conflicts if there is no epilogue in shared memory (with assumption that prolog shared memory does not have bank conflicts).

Maybe we could try to check how many TensorView objects in the fusion is in shared memory memory or if the producer for fusion output is in the shared memory?

Reply:
Make sense. I added check to make sure the correct number of shared memory tensors are generated.
```
    // There should be 3 shared memory tensors 2 for prologue and 1 for epilogue.
    int num_shared_mem_tensors = 0;
    for(const auto& tv : ir_utils::allTvs(&fusion)) {
      if(tv->getMemoryType() == MemoryType::Shared) {
        num_shared_mem_tensors++;
      }
    }
    TORCH_CHECK(num_shared_mem_tensors == 3, "Expected 3 shared memory tensors, got ", num_shared_mem_tensors);

```